test/functional/feature_fork_choice_forked_finalize_epoch.py

```diff
@@ -304,8 +285,7 @@ def test_fork_on_justified_epoch(self):
         # cleanup
         self.stop_node(node.index)
         self.stop_node(fork.index)
-        self.stop_node(finalizer1.index)
-        self.stop_node(finalizer2.index)
+        self.stop_node(finalizer.index)
```

Comment:
Any reason to stop the nodes manually?

Reply:
Basically, this file contains two tests. The first test uses the first three nodes and the second uses the last three.  And it's better to stop the nodes after each test so they could not interfere. That's my guess. We can also ask @kostyantyn

Reply:
Yes, this was the intention. To keep only those nodes that are needed for that sub-test.

Reply:
Although I were advocating such a way to write functional tests, now I see it is a mess. The less nodes started in every single unit of functional test is better. I panic anytime I see test launches 4+ nodes, because it's too way difficult to find corresponding logs when the test is failed. Also, having many suites in a single functional test file prevents from parallel tests execution.

Reply:
I think in general that if a test requires a different setup (i.e. dedicated nodes) then it should live in a different file.

Reply:
@frolosofsky since you mentioned few points, would like to comment on them individually.

>The less nodes started in every single unit of functional test is better

I agree that we need to have as little nodes as possible. 

>  I panic anytime I see test launches 4+ nodes, because it's too way difficult to find corresponding logs when the test is failed

There are scenarios that require more than 4 nodes to be able to simulate them. Currently, we have a hard limit of 8 so we don't go over it but if we lower that limit to 4, we would need to remove some tests

>, because it's too way difficult to find corresponding logs when the test is failed.

in debug log all lines are prefixed with the node index, you can grep by it.

> Also, having many suites in a single functional test file prevents from parallel tests execution.

We can extract sub-tests to the separate file if they don't reuse the same nodes. Sounds like a good idea.